client/state/comments/reducer.js

```diff
@@ -0,0 +1,176 @@
+ * Updates a specific state in the general state
+ * @param {Immutable.Map} state general state for all commentTargetId
+ * @param {Object} action action object
+ * @param {Number} action.siteId site identifier
+ * @param {Number} action.postId post identifier
+ * @param {Function|Object} updaterOrValue an updater function or a value
+ * @returns {Immutable.Map} new state
+ */
+function updateSpecificState( state, action, updaterOrValue ) {
+	const id = getCommentParentKey( action.siteId, action.postId );
+
+	if ( typeof updaterOrValue === 'function' ) {
+		return state.update( id, ( value ) => updaterOrValue( value ) );
+	}
+
+	return state.set( id, updaterOrValue );
+}
+
+/***
+ * Comments items reducer, stores a comments items Immutable.List per siteId, postId
+ * @param {Immutable.Map} state redux state
+ * @param {Object} action redux action
+ * @returns {Immutable.Map} new redux state
+ */
+export function items( state = Immutable.Map(), action ) {
+	switch ( action.type ) {
+		case COMMENTS_RECEIVE:
+			// create set of ids for faster lookup for filter later
+			const newIds = Immutable.Set( action.comments.map( comment => comment.ID ) );
+
+			// we prefer freshly retrieved data, so throw away old data
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				comments.filter( ( comment ) => ! newIds.has( comment.get( 'ID' ) ) )
+								.concat( Immutable.fromJS( action.comments ) )
+								.sort( ( a, b ) => new Date( b.get( 'date' ) ) - new Date( a.get( 'date' ) ) )
+			);
+
+		case COMMENTS_REMOVE:
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				comments.filter( ( comment ) => comment.get( 'ID' ) !== action.commentId )
+			);
+		case COMMENTS_LIKE:
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				listFindAndModify( comments,
+					( comment ) => comment.get( 'ID' ) === action.commentId,
+					getLikeSetter( { like: true } )
+				)
+			);
+		case COMMENTS_LIKE_UPDATE:
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				listFindAndModify( comments,
+					( comment ) => comment.get( 'ID' ) === action.commentId,
+					getLikeSetter( { like: action.iLike, likeCount: action.likeCount } )
+				)
+			);
+		case COMMENTS_UNLIKE:
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				listFindAndModify( comments,
+					( comment ) => comment.get( 'ID' ) === action.commentId,
+					getLikeSetter( { like: false } )
+				)
+			);
+		case COMMENTS_ERROR:
+			return updateSpecificState( state, action, ( comments = Immutable.List() ) =>
+				listFindAndModify( comments,
+					( comment ) => comment.get( 'ID' ) === action.commentId,
+					( comment ) => comment.set( 'placeholderState', PLACEHOLDER_STATE.ERROR )
+										.set( 'placeholderError', action.error )
+				)
+			);
+		case SERIALIZE:
+			return state.toJS();
+		case DESERIALIZE:
+			return Immutable.fromJS( state );
```

Comment:
If we want to persist this state tree, we'll want to add a JSON Schema for each subtree. We can also defer this work by temporarily choosing to not persist state.

```javascript
case SERIALIZE:
    return {};
case DESERIALIZE:
    return Immutable.Map();
```
The documentation update is still open, but if you're interested you can read some of the reasoning here: https://github.com/Automattic/wp-calypso/pull/3267